### PR TITLE
feat: hide non-visible variants to disallow selection

### PR DESCRIPTION
**Pull request by olehrakwix**
show option values that are present in visible variants to prevent user from selecting hidden product variant

https://github.com/user-attachments/assets/305850e7-1eca-42e0-a4f7-6f66ed01b51b

add out of stock choices indication:
image
image

### CODE REVIEW

File: src/components/product-option/product-option.tsx [hunk 1]
```diff
@@ -6,18 +6,23 @@ import { getChoiceValue } from '~/utils';
...
-export const ProductOption = ({ option, selectedChoice, error, onChange }: ProductOptionProps) => {
+export const ProductOption = ({ option, allowedValues, selectedChoice, error, onChange }: ProductOptionProps) => {
     const { name, optionType, choices } = option;
 
     if (name === undefined || choices === undefined) {
         return null;
     }
 
+    const allowedChoices = choices.filter(
+        (c) => allowedValues === undefined || allowedValues.has(getChoiceValue(optionType!, c)!)
```
Comment (VladyslavG) [inline src/components/product-option/product-option.tsx]: I don't understand this predicate. should we use `?` somewhere ?

File: src/components/product-option/product-option.tsx [hunk 2]
```diff
@@ -6,18 +6,23 @@ import { getChoiceValue } from '~/utils';
 
 export interface ProductOptionProps {
     option: products.ProductOption;
+    allowedValues: Set<string> | undefined;
```
Comment (yurii-ve) [inline src/components/product-option/product-option.tsx]: this complicates the component.. let's better make sure that option contains only allowed choices

File: app/routes/products.$productSlug/route.tsx [hunk 3]
```diff
@@ -67,6 +68,47 @@ export default function ProductDetailsPage() {
     const sku = getSKU(product, selectedChoices);
     const media = getMedia(product, selectedChoices);
 
+    /**
+     * Returns the set of allowed values for the specified option
+     * based on available product variants and other options choices.
+     */
+    function getAllowedOptionValues(
```
Comment (yurii-ve) [inline app/routes/products.$productSlug/route.tsx]: maybe `getValidOptionChoice(...): Choice[]`?

File: src/components/product-option/product-option.tsx (hunk 2 above)
Comment (yurii-ve) [reply src/components/product-option/product-option.tsx]: then this PR will change only 1 file 😄

Comment (olehrakwix) [reply src/components/product-option/product-option.tsx]: so then @VladyslavG would be very happy 😂

File: src/components/product-option/product-option.tsx (hunk 1 above)
Comment (VladyslavG) [reply src/components/product-option/product-option.tsx]: by the way first part of predicate can be extracted out

File: app/routes/products.$productSlug/route.tsx [hunk 4]
```diff
@@ -14,18 +15,55 @@ import { ProductOption } from '~/components/product-option/product-option';
...
+        [option.name]: undefined,
+    });
+
+    const choiceValue = getChoiceValue(option.optionType, choice);
+    return matchingVariants.some(
+        (variant) => variant.variant?.visible && variant.choices?.[option.name!] === choiceValue
+    );
+}
+
+function getActualOptions(
```
Comment (yurii-ve) [inline app/routes/products.$productSlug/route.tsx]: maybe move this to product utils and name it just `getProductOptions` with the comment what it does?

File: src/components/color-select/color-select.tsx [hunk 5]
```diff
@@ -18,12 +19,15 @@ export const ColorSelect = ({ options, selectedName, onChange, hasError }: Color
         <div className={styles.root}>
             {options.map((o) => (
                 <button
+                    disabled={!o.inStock}
```
Comment (yurii-ve) [inline src/components/color-select/color-select.tsx]: it shouldn't be disabled.. user should still be able to select it, to check out images, price, etc.

File: src/components/color-select/color-select.tsx [hunk 6]
```diff
@@ -18,12 +19,15 @@ export const ColorSelect = ({ options, selectedName, onChange, hasError }: Color
...
                     key={o.name}
                     className={classNames(styles.option, {
                         [styles.selected]: selectedName === o.name,
                         [styles.hasError]: hasError,
+                        [styles.outOfStock]: !o.inStock,
                     })}
                     onClick={() => onChange(o.name!)}
+                    title={!o.inStock ? 'Out of stock' : undefined}
```
Comment (yurii-ve) [inline src/components/color-select/color-select.tsx]: no need, `disabled` is removed

File: src/components/select/select.tsx [hunk 7]
```diff
@@ -29,8 +30,9 @@ export const Select = ({ options, value, onChange, placeholder, hasError }: Sele
                 ) : null}
 
                 {options.map((o) => (
-                    <option key={o.value} value={o.value}>
+                    <option key={o.value} value={o.value} disabled={!o.inStock}>
```
Comment (yurii-ve) [inline src/components/select/select.tsx]: not disabled as well

File: src/components/color-select/color-select.tsx (hunk 5 above)
Comment (olehrakwix) [reply src/components/color-select/color-select.tsx]: fixed

File: src/utils/product-utils.ts [hunk 8]
```diff
@@ -88,3 +112,43 @@ export function getMedia(
     const selectedChoiceWithMedia = Object.values(selectedChoices).find((c) => c?.media?.mainMedia !== undefined);
     return selectedChoiceWithMedia?.media ?? product.media;
 }
+
+export function getProductOptions(
```
Comment (yurii-ve) [inline src/utils/product-utils.ts:118]: let's add a comment that describes what this function does

File: src/utils/product-utils.ts [hunk 9]
```diff
@@ -43,12 +43,30 @@ export function getSKU(
...
 ): wixStoresProducts.Variant | undefined {
     const selectedChoiceValues = selectedChoicesToVariantChoices(product, selectedChoices);
     return product.variants?.find((variant) => deepEqual(variant.choices, selectedChoiceValues));
 }
 
+export function getMatchingVariants(
+    product: Product | SerializeFrom<Product>,
+    selectedChoices: Record<string, wixStoresProducts.Choice | undefined>
+): wixStoresProducts.Variant[] {
+    const selectedChoiceValues = selectedChoicesToVariantChoices(product, selectedChoices, {
```
Comment (yurii-ve) [inline src/utils/product-utils.ts]: let's just filter out attributes with `undefined` values here and not introduce the `option` parameter to the `selectedChoicesToVariantChoices`.

File: src/utils/product-utils.ts [hunk 10]
```diff
@@ -88,3 +112,43 @@ export function getMedia(
...
+    product: Product | SerializeFrom<Product>
+): Pick<wixStoresProducts.Choice, 'visible' | 'inStock'> {
+    if (!product.manageVariants || !option.name || !option.optionType) {
+        return {
+            visible: choice.visible,
+            inStock: choice.inStock,
+        };
+    }
+
+    // Get variants matching all other selected choices and target choice
```
Comment (yurii-ve) [inline src/utils/product-utils.ts:146]: nice 👍

File: src/utils/product-utils.ts [hunk 11]
```diff
@@ -43,12 +43,30 @@ export function getSKU(
...
 export function getSelectedVariant(
     product: Product | SerializeFrom<Product>,
-    selectedChoices: Record<string, wixStoresProducts.Choice | undefined> = {}
+    selectedChoices: Record<string, wixStoresProducts.Choice | undefined>
 ): wixStoresProducts.Variant | undefined {
     const selectedChoiceValues = selectedChoicesToVariantChoices(product, selectedChoices);
     return product.variants?.find((variant) => deepEqual(variant.choices, selectedChoiceValues));
 }
 
+export function getMatchingVariants(
```
Comment (yurii-ve) [inline src/utils/product-utils.ts]: let's not export this function as it's not used anywhere else

**Review by yurii-ve — APPROVED** [review]

**Pull request merged**